### DISCUSSION TITLE
New Host Applications

### DISCUSSION THREAD

**Discussion by dcrawl** [bnpr/Malt#389] [Development]
I'm sure this is beyond my skillset but it's tempting to try. :)   I'd like to have the NPR rendering working for another host application that only supports basic NPR.  How difficult is it to make Malt compatible with other applications that support Python integration?  Any advice on where to start on the Malt side of things?

Thanks.

**Comment by pragma37** [comment]
The Malt and Bridge packages are designed to be entirely agnostic from the host, so it should be possible to use them as-is.

The steps a host should follow would be:

- Create a [Bridge instance](https://github.com/bnpr/Malt/blob/Development/Bridge/Client_API.py).
- Read the [Parameters](https://github.com/bnpr/Malt/blob/Development/Malt/PipelineParameters.py) that the pipeline expects (Bridge.get_parameters()) and make sure they are editable in the host UI.
- At render time:
    - Make sure all material have been compiled (Bridge.compile_materials).
    - Make sure meshes (Bridge.load_mesh()), textures (Bridge.load_texture()) and gradients (Bridge.load_gradient()) have been loaded.
    - Populate a [Scene](https://github.com/bnpr/Malt/blob/Development/Malt/Scene.py), using [Proxys](https://github.com/bnpr/Malt/blob/Development/Bridge/Proxys.py) for shaders, meshes and textures.
    - Send the scene to your Bridge instance (Bridge.render()) and get back the result (Bridge.render_result()).
    
Getting something like the [MiniPipeline](https://github.com/bnpr/Malt/blob/Development/Malt/Pipelines/MiniPipeline/MiniPipeline.py) working shouldn't take a terrible amount of work.

You could even start by skipping the parameters UI, for example, and simply send back the default values.
The same for Materials, you could ignore shader parameters and let Malt just use the default values.

The hairy part would be supporting the node system.
Luckily, you can do a lot of stuff in Malt without nodes.
